File: src/livekit_agent_simulator/caller_nudge.py
"""Nudge the simulated caller to speak after the agent greets first."""

from __future__ import annotations

import asyncio
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from .callers.gemini import GeminiCallerBridge
    from .livekit.observer import Observer
    from .logging.event_writer import EventWriter

# Language-neutral: persona / scenario locale already constrain speech language.
#
# NOTE: this nudge must NOT be spoken aloud. The old implementation injected
# the text as a role=user conversation item + response.create, so the OpenAI
# caller literally READ the instruction out loud ("has finished greeting you.
# Respond now in the language of your persona") — the agent then heard it as a
# real caller utterance and the live transcript diverged from the voice. The
# caller's persona brief already tells it how to respond, so the correct nudge
# is a NON-AUDIBLE "please reply to the last agent turn" activation
# (nudge_freestyle_answer), not an audible spoken line.
AGENT_GREETED_NUDGE = (
    "(The agent has finished greeting you. Respond now in the language of your persona.)"
)


async def nudge_caller_after_agent_greeting(
    observer: "Observer",
    bridge: "GeminiCallerBridge",
    writer: "EventWriter",
    *,
    first_speaker: str,
    debounce_s: float = 1.0,
    poll_s: float = 0.15,

    silent_mode: bool = False,
) -> None:
    """When first_speaker is agent, persona-only runs stall without a text bootstrap.

    Activates the caller via ``nudge_freestyle_answer`` (commit the agent audio
    + response.create) so the caller replies naturally to the greeting — the
    nudge text itself is NEVER spoken into the room.

    Silent mode (Coval): never nudge — unresponsive caller must stay mute.
    """
    if first_speaker != "agent":
        return
    if silent_mode or getattr(bridge, "_silent_mode", False) is True:
        writer.emit(
            "sim.agent_greeted_nudge_skipped",
            spec={"reason": "silent_mode"},

            source="sim",
            include_dialogue=False,
        )
        return

    nudged = False
    while not bridge.end_call.is_set():
        if nudged:
            return
        if observer.agent_has_spoken and not observer.user_has_spoken:
            await asyncio.sleep(debounce_s)
            if bridge.end_call.is_set() or observer.user_has_spoken:
                return
            try:
                # Non-audible activation: do NOT speak the nudge text into the
                # room (that leaks the instruction as fake caller audio).
                await bridge.nudge_freestyle_answer(agent_hint=AGENT_GREETED_NUDGE)
                writer.emit(
                    "sim.agent_greeted_nudge",
                    spec={"debounce_s": debounce_s, "audible": False},
                    source="sim",
                    include_dialogue=False,
                )
                nudged = True
                return
            except RuntimeError:
                await asyncio.sleep(poll_s)
                continue
        await asyncio.sleep(poll_s)

File: src/livekit_agent_simulator_rust/crates/lks-livekit/src/caller_nudge.rs
//! Caller nudge (port of `caller_nudge.py`) — when first_speaker is agent,
//! persona-only runs stall without a text bootstrap. This activates the caller
//! via response.create (the nudge text is NEVER spoken into the room).

use std::sync::Arc;
use tokio::sync::{broadcast, Mutex};

use lks_core::logging::event::EventWriter;

pub const AGENT_GREETED_NUDGE: &str =
    "(The agent has finished greeting you. Respond now in the language of your persona.)";

/// Nudge the caller after the agent's greeting (non-audible activation).
#[allow(clippy::too_many_arguments)]
pub async fn nudge_caller_after_agent_greeting(
    writer: Arc<Mutex<EventWriter>>,
    end_call: broadcast::Receiver<()>,
    on_nudge: impl Fn(&str) -> Result<(), String>,
    agent_has_spoken: impl Fn() -> bool + Send + 'static,
    user_has_spoken: impl Fn() -> bool + Send + 'static,
    first_speaker: &str,
    silent_mode: bool,
    debounce_s: f64,
    poll_s: f64,
) {
    if first_speaker != "agent" {
        return;
    }
    if silent_mode {
        let mut w = writer.lock().await;
        w.emit(
            "sim.agent_greeted_nudge_skipped",
            Some(
                &serde_json::json!({"reason": "silent_mode"})
                    .as_object()
                    .cloned()
                    .unwrap_or_default(),
            ),
            "sim",
            None,
            None,
            false,
            None,
        );
        return;
    }
    let mut end_call = end_call;
    let mut nudged = false;
    while !nudged && end_call.try_recv().is_err() {
        if agent_has_spoken() && !user_has_spoken() {
            tokio::time::sleep(std::time::Duration::from_secs_f64(debounce_s)).await;
            if end_call.try_recv().is_ok() || user_has_spoken() {
                return;
            }
            match on_nudge(AGENT_GREETED_NUDGE) {
                Ok(()) => {
                    let mut w = writer.lock().await;
                    w.emit(
                        "sim.agent_greeted_nudge",
                        Some(
                            &serde_json::json!({"debounce_s": debounce_s, "audible": false})
                                .as_object()
                                .cloned()
                                .unwrap_or_default(),
                        ),
                        "sim",
                        None,
                        None,
                        false,
                        None,
                    );
                    nudged = true;
                }
                Err(_) => {
                    tokio::time::sleep(std::time::Duration::from_secs_f64(poll_s)).await;
                }
            }
        } else {
            tokio::time::sleep(std::time::Duration::from_secs_f64(poll_s)).await;
        }
    }
}
